File: timemanager-web/src/types/preferences.ts
// Type definitions for time-related values
export type TimeString = string; // Format: "HH:mm"
export type DayOfWeek = "monday" | "tuesday" | "wednesday" | "thursday" | "friday";
export type Language = "en" | "zh";

// Validation utilities
export const isValidTimeString = (time: string): boolean => {
  const timeRegex = /^([0-1][0-9]|2[0-3]):[0-5][0-9]$/;
  return timeRegex.test(time);
};

export const isValidDayOfWeek = (day: string): day is DayOfWeek => {
  return ["monday", "tuesday", "wednesday", "thursday", "friday"].includes(day.toLowerCase());
};

export interface WorkingHours {
  start: TimeString;
  end: TimeString;
}

export interface NotificationPreferences {
  email: boolean;
  push: boolean;
  wechat: boolean;
}

export interface CalendarSync {
  google: boolean;
  outlook: boolean;
  apple: boolean;
}

export interface UserPreferences {
  user_id?: string;
  timezone: string;
  language: Language;
  notification_preferences: NotificationPreferences;
  working_hours: { [key in DayOfWeek]: WorkingHours };
  calendar_sync: CalendarSync;
}

// Helper function to get default preferences
export const getDefaultPreferences = (): UserPreferences => ({
  timezone: "UTC",
  language: "en",
  notification_preferences: {
    email: true,
    push: true,
    wechat: true
  },
  working_hours: {
    monday: { start: "09:00", end: "17:00" },
    tuesday: { start: "09:00", end: "17:00" },
    wednesday: { start: "09:00", end: "17:00" },
    thursday: { start: "09:00", end: "17:00" },
    friday: { start: "09:00", end: "17:00" }
  },
  calendar_sync: {
    google: false,
    outlook: false,
    apple: false
  }
});

File: timemanager/app/models/preferences.py
from typing import Dict, Optional
from pydantic import BaseModel, Field, validator
from datetime import time

class WorkingHours(BaseModel):
    start: str = Field(..., description="Start time in HH:mm format")
    end: str = Field(..., description="End time in HH:mm format")

    @validator("start", "end")
    def validate_time_format(cls, v):
        try:
            hour, minute = map(int, v.split(":"))
            if not (0 <= hour < 24 and 0 <= minute < 60):
                raise ValueError
            return v
        except:
            raise ValueError("Time must be in HH:mm format")

class NotificationPreferences(BaseModel):
    email: bool = True
    push: bool = True
    wechat: bool = True

class CalendarSync(BaseModel):
    google: bool = False
    outlook: bool = False
    apple: bool = False

class UserPreferences(BaseModel):
    user_id: Optional[str] = None
    timezone: str = "UTC"
    language: str = "en"
    notification_preferences: NotificationPreferences = Field(
        default_factory=NotificationPreferences
    )
    working_hours: Dict[str, WorkingHours] = Field(
        default_factory=lambda: {
            "monday": WorkingHours(start="09:00", end="17:00"),
            "tuesday": WorkingHours(start="09:00", end="17:00"),
            "wednesday": WorkingHours(start="09:00", end="17:00"),
            "thursday": WorkingHours(start="09:00", end="17:00"),
            "friday": WorkingHours(start="09:00", end="17:00")
        }
    )
    calendar_sync: CalendarSync = Field(default_factory=CalendarSync)

    @validator("language")
    def validate_language(cls, v):
        if v not in ["en", "zh"]:
            raise ValueError("Language must be either 'en' or 'zh'")
        return v
